File: app/src/main/java/com/meditrack/app/receivers/AlarmReceiver.java
package com.meditrack.app.receivers;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import com.meditrack.app.data.AppExecutors;
import com.meditrack.app.data.IntakeLog;
import com.meditrack.app.data.Medication;
import com.meditrack.app.data.PrefsManager;
import com.meditrack.app.data.Schedule;
import com.meditrack.app.db.DatabaseHelper;
import com.meditrack.app.db.MedicationDao;
import com.meditrack.app.services.AlarmScheduler;
import com.meditrack.app.services.NotificationHelper;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class AlarmReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        int medicationId = intent.getIntExtra("MEDICATION_ID", -1);
        String medicationName = intent.getStringExtra("MEDICATION_NAME");
        String dosage = intent.getStringExtra("DOSAGE");
        int scheduleId = intent.getIntExtra("SCHEDULE_ID", -1);

        if (medicationId == -1 || scheduleId == -1) {
            return;
        }

        AppExecutors.getInstance().diskIO(() -> {
            MedicationDao dao = new MedicationDao(DatabaseHelper.getInstance(context));
            IntakeLog log = new IntakeLog();
            log.setMedicationId(medicationId);
            log.setScheduledDatetime(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss",
                    Locale.getDefault()).format(new Date()));
            log.setTaken(false);
            dao.insertIntakeLog(log);
        });

        NotificationHelper.showMedicationReminder(context,
                medicationId,
                medicationName != null ? medicationName : "",
                dosage != null ? dosage : "",
                scheduleId);

        scheduleMissedDoseCheck(context, medicationId, scheduleId);

        AppExecutors.getInstance().diskIO(() -> {
            MedicationDao dao = new MedicationDao(DatabaseHelper.getInstance(context));
            Schedule schedule = null;
            for (Schedule s : dao.getAllSchedules()) {
                if (s.getId() == scheduleId) {
                    schedule = s;
                    break;
                }
            }
            if (schedule != null) {
                Medication med = dao.getMedicationById(medicationId);
                if (med != null && med.isActive()) {
                    AlarmScheduler.scheduleAlarm(context.getApplicationContext(), schedule, med);
                }
            }
        });
    }

    private void scheduleMissedDoseCheck(Context context, int medicationId, int scheduleId) {
        int checkDelay = PrefsManager.getReminderMinutes(context) * 2;
        long checkAt = System.currentTimeMillis() + checkDelay * 60 * 1000L;

        Intent checkIntent = new Intent(context, MissedDoseReceiver.class);
        checkIntent.putExtra("MEDICATION_ID", medicationId);
        checkIntent.putExtra("SCHEDULE_ID", scheduleId);

        PendingIntent checkPI = PendingIntent.getBroadcast(context,
                scheduleId + 30000,
                checkIntent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        AlarmManager am = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        if (am == null) {
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (am.canScheduleExactAlarms()) {
                am.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, checkAt, checkPI);
            }
        } else {
            am.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, checkAt, checkPI);
        }
    }
}

File: app/src/main/java/com/meditrack/app/receivers/AlarmReceiver.kt
package com.meditrack.app.receivers

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.meditrack.app.database.DatabaseHelper
import com.meditrack.app.database.IntakeLogDAO
import com.meditrack.app.models.IntakeLog
import com.meditrack.app.notifications.NotificationHelper
import com.meditrack.app.utils.AlarmScheduler
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * AlarmReceiver - מקבל אירועי שעון מעורר ופעולות מהתראה
 * אחראית: רע'ד
 */
class AlarmReceiver : BroadcastReceiver() {

    companion object {
        const val EXTRA_MED_ID = "med_id"
        const val EXTRA_MED_NAME = "med_name"
        const val EXTRA_MED_DOSAGE = "med_dosage"
        const val EXTRA_SCHEDULE_ID = "schedule_id"
        const val EXTRA_LOG_ID = "log_id"
    }

    private val executor: ExecutorService = Executors.newSingleThreadExecutor()

    override fun onReceive(context: Context, intent: Intent) {
        when (intent.action) {
            // התראה חדשה - הגיע זמן לקחת תרופה
            Intent.ACTION_BOOT_COMPLETED -> {
                // לאחר אתחול המכשיר - קבע מחדש את כל האזעקות
                rescheduleAllAlarms(context)
            }

            NotificationHelper.ACTION_TAKEN -> {
                val logId = intent.getLongExtra(EXTRA_LOG_ID, -1L)
                if (logId != -1L) markAsTaken(context, logId)
            }

            NotificationHelper.ACTION_SNOOZE -> {
                val logId = intent.getLongExtra(EXTRA_LOG_ID, -1L)
                if (logId != -1L) snoozeReminder(context, logId)
            }

            else -> {
                // אזעקה רגילה - הצג התראה
                val medId = intent.getLongExtra(EXTRA_MED_ID, -1L)
                val medName = intent.getStringExtra(EXTRA_MED_NAME) ?: return
                val medDosage = intent.getStringExtra(EXTRA_MED_DOSAGE) ?: ""
                val scheduleId = intent.getLongExtra(EXTRA_SCHEDULE_ID, -1L)

                if (medId != -1L) fireReminder(context, medId, medName, medDosage, scheduleId)
            }
        }
    }

    private fun fireReminder(
        context: Context,
        medId: Long,
        medName: String,
        dosage: String,
        scheduleId: Long
    ) {
        executor.execute {
            val dbHelper = DatabaseHelper(context)
            val logDAO = IntakeLogDAO(dbHelper)

            // צור רשומת יומן
            val logId = logDAO.insert(
                IntakeLog(
                    medicationId = medId,
                    scheduledTime = System.currentTimeMillis(),
                    status = IntakeLog.Status.PENDING
                )
            )

            NotificationHelper.showMedicationReminder(context, logId, medName, dosage)
        }
    }

    private fun markAsTaken(context: Context, logId: Long) {
        executor.execute {
            val dbHelper = DatabaseHelper(context)
            val logDAO = IntakeLogDAO(dbHelper)
            logDAO.updateStatus(logId, IntakeLog.Status.TAKEN, System.currentTimeMillis())
            NotificationHelper.cancelNotification(context, logId)
        }
    }

    private fun snoozeReminder(context: Context, logId: Long) {
        executor.execute {
            val dbHelper = DatabaseHelper(context)
            val logDAO = IntakeLogDAO(dbHelper)
            logDAO.updateStatus(logId, IntakeLog.Status.SNOOZED)
            NotificationHelper.cancelNotification(context, logId)

            // קבע מחדש עוד 10 דקות
            AlarmScheduler.scheduleSnooze(context, logId, delayMinutes = 10)
        }
    }

    private fun rescheduleAllAlarms(context: Context) {
        executor.execute {
            AlarmScheduler.rescheduleAll(context)
        }
    }
}
